Taylor and Francis Group.js

```diff
@@ -0,0 +1,287 @@
+			if (items) ZU.processDocuments(Object.keys(items), scrape);
+		});
+	}
+	else {
+		scrape(doc, url, itemType);
+	}
+}
+
+function scrape(doc, url, itemType) {
+	if (!itemType) {
+		itemType = detectWeb(doc, url);
+	}
+
+	let item = new Zotero.Item(itemType);
+	item.title = text(doc, 'h1.product-banner-title');
+	item.libraryCatalog = '';
+	item.extra = '';
+	let subtitle = text(doc, '.product-banner-subtitle');
+	if (subtitle) {
+		item.title += ": " + subtitle;
+	}
+	if (itemType == 'bookSection') {
+		item.bookTitle = text(doc, 'a.product-book-link');
+	}
+	
+	for (let creator of text(doc, '.product-banner-author-name').split(", ")) {
+		let creatorType = text(doc, '.product-banner-author').includes('Edited')
+			? 'editor'
+			: 'author';
+		item.creators.push(ZU.cleanAuthor(creator, creatorType, false));
+	}
+	
+	for (let row of doc.querySelectorAll('product-more-details .display-row')) {
+		let label = text(row, 'span:first-child');
+		let value = text(row, 'span:last-child');
+
+		let field = FIELDS[label];
+		if (field) {
+			if (field == 'extra') {
+				if (label == "DOI") {
+					value = value.match(/(?:doi\.org\/)?([^/]+\/[^/]+)/)[1];
+				}
+				item.extra += `${label}: ${value}\n`;
+			}
+			else {
+				item[field] = value;
+			}
+		}
+	}
+	
+	if (doc.cookie) {
+		let apiId = doc.getElementById('pyramid-website-ssr-state').textContent
+			.match(/&q;_id&q;:&q;([^&]+)&q;/)[1];
+		let token = doc.cookie
+			.split('; ')
+			.find(row => row.startsWith('_token='))
+			.split('=')[1];
+		ZU.doGet(`https://api.taylorfrancis.com/v4/content/${apiId}?apiVersion=4.0.1&filenamePrefix=${item.ISBN}&render=false`,
+			function (respText, xhr) {
+				// the request might fail if we don't have PDF access
+				// but that's alright, everything else worked
+				// so we treat every status as a success (last arg to doGet)
+				// but only try to fetch the attachment if we got 200 OK
+				
+				if (xhr.status == 200) {
+					let resp = JSON.parse(respText);
+					for (let contentObject of resp) {
+						if (contentObject.type == 'webpdf'
+							|| contentObject.type == 'chapterpdf') {
+							item.attachments.push({
+								title: "Full Text PDF",
+								mimeType: "application/pdf",
+								url: contentObject.location
+							});
+							break;
+						}
+					}
+				}
+				
+				item.complete();
+			}, null, null, {
+				Authorization: 'idtoken ' + token,
+				Referrer: 'https://www.taylorfrancis.com/'
+			}, false);
+	}
+	else {
+		item.complete();
+	}
+}
+
+/** BEGIN TEST CASES **/
+var testCases = [
+	{
+		"type": "web",
+		"url": "https://www.taylorfrancis.com/books/edit/10.1201/9781315120621/lubricant-additives-leslie-rudnick",
+		"items": [
+			{
+				"itemType": "book",
+				"title": "Lubricant Additives: Chemistry and Applications",
+				"creators": [
+					{
+						"firstName": "Leslie R.",
+						"lastName": "Rudnick",
+						"creatorType": "editor"
+					}
+				],
+				"date": "29 June 2017",
```

Comment:
Our date handling/localization is complicated and imperfect, but we generally use `strToISO()` to save parse dates into YYYY-MM-DD.

(Basically, `strToDate()`, which that uses, can parse English month names in any user locale, so that will work here and the user won't have to see English month names and British date order. The ultimate plan is to display cleanly parsed Date values in the user's own locale, but we haven't gotten around to that.)